dpll_mon.c

```diff
@@ -0,0 +1,991 @@
+		STAILQ_REMOVE_HEAD(&dm->pins, list);
+		pin_destroy(pin);
+	}
+	free(dm);
+}
+
+int dpll_mon_init(struct dpll_mon *dm)
+{
+	int ret;
+
+	if (dm->state != DPLL_MON_STATE_CREATED) {
+		pr_debug("%s wrong dpll monitor state for %s device",
+			 __func__, dm->name);
+		return -ENODEV;
+	}
+
+	dm->mon_sk = nl_dpll_mon_socket_create(dpll_mon_ntf_recv, dm);
+	if (dm->mon_sk) {
+		pr_debug("monitor socket created");
+	} else {
+		pr_debug("failed to create monitor socket");
+		return -EINVAL;
+	}
+
+	dm->dev_sk = nl_dpll_sk_create();
+	if (dm->dev_sk) {
+		pr_debug("dev socket created for %s device", dm->name);
+	} else {
+		pr_err("dev socket failed for %s device", dm->name);
+		return -EINVAL;
+	}
+	dm->family = nl_dpll_family_resolve(dm->dev_sk);
+	if (dm->family < 0)
+		pr_err("nl dpll family not found for %s device", dm->name);
+
+	dm->rt_sk = nl_rt_sk_create();
+	if (!dm->rt_sk) {
+		pr_err("rt socket failed for %s device", dm->name);
+		return -EINVAL;
+	}
+	STAILQ_INIT(&dm->pins);
+	dpll_mon_state_set(dm, DPLL_MON_STATE_INIT_READY);
+	ret = dpll_mon_thread_create(dm);
+	if (ret) {
+		dpll_mon_destroy(dm);
+		return ret;
+	}
+
+	return 0;
+}
+
+int dpll_mon_mode_get(struct dpll_mon *dm, enum dpll_mode *mode)
+{
+	*mode = dm->dpll_mode;
+
+	return 0;
+}
+
+int dpll_mon_lock_state_get(struct dpll_mon *dm, enum eec_state *state)
+{
+	*state = lock_status_to_state[dm->lock_status];
+
+	return 0;
+}
+
+struct dpll_mon_pin
+*dpll_mon_add_pin(struct dpll_mon *dm, const char *board_label,
+		  const char *panel_label, const char *package_label,
+		  enum dpll_pin_type type)
+{
+	struct dpll_mon_pin *pin = pin_create();
+	int ret = 0;
+
+	if (!pin)
+		return NULL;
+	pin->board_label = board_label;
+	pin->package_label = package_label;
+	pin->panel_label = panel_label;
+	pin->type = type;
+	pin->valid = PIN_VALID;
+	STAILQ_INSERT_TAIL(&dm->pins, pin, list);
+	pr_debug_pin("adding valid pin", pin);
+	if (ret)
+		pr_debug_pin("failed to send get id request", pin);
+	return pin;
+}
+
+struct dpll_mon_pin
+*dpll_mon_add_port_pin(struct dpll_mon *dm, const char *ifname)
+{
+	struct dpll_mon_pin *pin = pin_create();
+	int fd;
+
+	if (!pin)
+		return NULL;
+	pin->ifname = ifname;
+	fd = socket(AF_PACKET, SOCK_RAW | SOCK_NONBLOCK, PF_UNIX);
+	if (fd < 0) {
+		pr_err("create socket for %s failed: %m", ifname);
+		if (fd == -EPERM)
+			pr_err("try running as root");
+		goto error;
+	}
+	pin->if_index = sk_interface_index(fd, ifname);
+	if (pin->if_index < 0) {
+		pr_err("sk_interface_index for %s failed: %m", ifname);
+		close(fd);
+		goto error;
+	}
+	close(fd);
+	pin->type = DPLL_PIN_TYPE_SYNCE_ETH_PORT;
+	pin->valid = PIN_VALID;
+	STAILQ_INSERT_TAIL(&dm->pins, pin, list);
+	pr_debug_pin("adding valid pin", pin);
+
+	return pin;
+error:
+	free(pin);
+	return NULL;
+}
+
+int dpll_mon_pin_is_active(struct dpll_mon *dm, struct dpll_mon_pin *pin)
+{
+	struct dpll_mon_pin *parent;
+	struct parent_pin *pp;
+
+	if (pin->muxed) {
```

Comment:
Similar thing here, not a big thing but if we reversed the logic we would get more readable code - what do you think?
```
if (!pin->muxed)
	return pin->state == DPLL_PIN_STATE_CONNECTED;

STAILQ_FOREACH(pp, &pin->parents, list)
	if (pp->state == DPLL_PIN_STATE_CONNECTED) {
		parent = find_pin(dm, pp->id);
		if (parent->state == DPLL_PIN_STATE_CONNECTED)
			return 1;
	}
return 0;
```

Reply:
yep, will fix!